### PR TITLE
Fix match on vec items: match on vec[..]

### LINKED ISSUE #5551: match_on_vec_items: false positive on `match vec[..]`
[Sample program](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1b2c6a87fdacb613af670b53cf094bac):
```rust
fn main() {
    let x = vec![1,2,3];
    match x[..] {
        [1, 2, 3] => println!("123"),
        _ => println!("not 123"),
    }
}
```

Clippy output:
```
error: indexing into a vector may panic
 --> src/main.rs:3:11
  |
3 |     match x[..] {
  |           ^^^^^ help: try this: `x.get(..)`
  |
  = note: `#[deny(clippy::match_on_vec_items)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_on_vec_items

error: aborting due to previous error
```

However, this use of indexing is infallible.

(note that I'm silencing this lint globally anyway because I don't consider it useful)

### PR DESCRIPTION
**Pull request by CrazyRoka**
- Added new tests
- Fixed false positive when matching on full range, which will never panic

Closes #5551 
changelog: fix match_on_vec_items when matching full range

### CODE REVIEW

File: clippy_lints/src/match_on_vec_items.rs [hunk 1]
```diff
@@ -87,3 +86,15 @@ fn is_vec_indexing<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'tcx>)
...
+fn is_vector(cx: &LateContext<'_, '_>, expr: &Expr<'_>) -> bool {
+    let ty = cx.tables.expr_ty(expr);
+    let ty = walk_ptrs_ty(ty);
+    is_type_diagnostic_item(cx, ty, sym!(vec_type))
+}
+
+fn is_full_range(cx: &LateContext<'_, '_>, expr: &Expr<'_>) -> bool {
+    let ty = cx.tables.expr_ty(expr);
+    let ty = walk_ptrs_ty(ty);
+    match_type(cx, ty, &["core", "ops", "range", "RangeFull"])
```
Comment (phansch) [inline clippy_lints/src/match_on_vec_items.rs:99]: Does this work as well?

```suggestion
    match_type(cx, ty, &utils::paths::RANGE_FULL)
```

File: tests/ui/match_on_vec_items.rs [hunk 2]
```diff
@@ -120,11 +120,33 @@ fn match_with_array() {
     }
 }
 
+fn match_with_endless_range() {
+    let arr = vec![0, 1, 2, 3];
+    let range = ..;
```
Comment (phansch) [inline tests/ui/match_on_vec_items.rs:125]: :exploding_head:

File: clippy_lints/src/match_on_vec_items.rs (hunk 1 above)
Comment (CrazyRoka) [reply clippy_lints/src/match_on_vec_items.rs:99]: No, this doesn't work
```Rust
pub const RANGE_FULL: [&str; 3] = ["core", "ops", "RangeFull"];
```

Comment (flip1995) [reply clippy_lints/src/match_on_vec_items.rs:99]: Then the path in `util::paths` is wrong. Can you change that and use it here afterwards?

Comment (CrazyRoka) [reply clippy_lints/src/match_on_vec_items.rs]: Updated it

**Pull request merged**